### PR TITLE
feat: add a simple way to convert Hugging Face model to GGUF

### PR DESCRIPTION
**Pull request by Helloyunho**
Note: Join Jan's Discord server and go to `roadmap` -> `Jan can convert Hugging Face models to GGUF` for more frequent updates.

## Describe Your Changes

- Add a way to convert Hugging Face models to GGUF in Jan
- (Hopefully) make decent metadata from Hugging Face's model card

## Fixes Issues

- None

## Self Checklist

- [ ] Added relevant comments, esp in complex areas
  - TODO
- [ ] Updated docs (for bug fixes / features)
  - TODO
- [ ] Created issues for follow-up changes or refactoring needed

## TODO (yeah I just made this up)

- All of the above
- Have a better UI (we really need it, the current UI that I made is um, filthy)
- Alert when the user doesn't have Python (check below)
- Build and bundle llama.cpp's quantize binary(like we do for nitro)

## Things to know (again I added this)

- User should have Python installed locally
  - Long explanation: This change uses llama.cpp's convert script to convert models, and since the script is written in Python, we need it to run the script.
- For easy debugging, it doesn't remove the model directory when the conversion has failed. Later we should uncomment it before publishing.

### CODE REVIEW

**Comment by 0xSage** [comment]
@louis-jan anything we can do to help Yunho get this merged? Thanks everyone

**Comment by louis-jan** [comment]
> @louis-jan anything we can do to help Yunho get this merged? Thanks everyone


> @louis-jan anything we can do to help Yunho get this merged? Thanks everyone

Sure.
Awesome work @Helloyunho. We will review and proceed with this shortly. Could you please help rebase latest dev in the meantime? Thanks.

**Comment by Helloyunho** [comment]
> > @louis-jan anything we can do to help Yunho get this merged? Thanks everyone
>
> Sure.
> 
> Awesome work @Helloyunho. We will review and proceed with this shortly. Could you please help rebase latest dev in the meantime?
> 
> Thanks.

Sure I'll get those things done

Also note that this PR sure needs some improvements from Jan's team, like the UI and bundling llama.cpp's quantize binary and so on (check the original comment)

I can do it but I'm not sure if I can do this in time(I have only a couple of weeks to code now)

**Comment by louis-jan** [comment]
> > > @louis-jan anything we can do to help Yunho get this merged? Thanks everyone
> > 
> > 
> > Sure.
> > Awesome work @Helloyunho. We will review and proceed with this shortly. Could you please help rebase latest dev in the meantime?
> > Thanks.
> 
> Sure I'll get those things done
> 
> Also note that this PR sure needs some improvements from Jan's team, like the UI and bundling llama.cpp's quantize binary and so on (check the original comment)
> 
> I can do it but I'm not sure if I can do this in time(I have only a couple of weeks to code now)



> > > @louis-jan anything we can do to help Yunho get this merged? Thanks everyone
> > 
> > 
> > Sure.
> > Awesome work @Helloyunho. We will review and proceed with this shortly. Could you please help rebase latest dev in the meantime?
> > Thanks.
> 
> Sure I'll get those things done
> 
> Also note that this PR sure needs some improvements from Jan's team, like the UI and bundling llama.cpp's quantize binary and so on (check the original comment)
> 
> I can do it but I'm not sure if I can do this in time(I have only a couple of weeks to code now)

Sweet, I appreciate your contributions and support. Thanks for your notes, we could do it together.

File: extensions/huggingface-extension/package.json [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  "version": "1.0.0",
+  "description": "Hugging Face extension for converting HF models to GGUF",
+  "main": "dist/index.js",
+  "node": "dist/node/index.cjs.js",
+  "author": "Jan <service@jan.ai>",
+  "license": "AGPL-3.0",
+  "scripts": {
+    "build": "tsc --module commonjs && rollup -c rollup.config.ts --configPlugin @rollup/plugin-typescript --bundleConfigAsCjs",
+    "download:llama": "run-script-os",
+    "download:llama:linux": "LLAMA_CPP_VERSION=$(cat ./scripts/version.txt) && download https://github.com/ggerganov/llama.cpp/archive/refs/tags/${LLAMA_CPP_VERSION}.tar.gz -o . --filename ./scripts/llama.cpp.tar.gz && tar -xzf ./scripts/llama.cpp.tar.gz '*/convert.py' '*/convert-hf-to-gguf.py' '*/gguf-py' && cpx \"./llama.cpp-$LLAMA_CPP_VERSION/**\" \"scripts\" && rimraf \"./scripts/llama.cpp.tar.gz\" && rimraf \"./llama.cpp-$LLAMA_CPP_VERSION\"",
```
Comment (hiento09) [inline extensions/huggingface-extension/package.json:12]:
```suggestion
    "download:llama:linux": "LLAMA_CPP_VERSION=$(cat ./scripts/version.txt) && download https://github.com/ggerganov/llama.cpp/archive/refs/tags/${LLAMA_CPP_VERSION}.tar.gz -o . --filename ./scripts/llama.cpp.tar.gz && tar -xzf ./scripts/llama.cpp.tar.gz --wildcards '*/convert.py' '*/convert-hf-to-gguf.py' '*/gguf-py' && cpx \"./llama.cpp-$LLAMA_CPP_VERSION/**\" \"scripts\" && rimraf \"./scripts/llama.cpp.tar.gz\" && rimraf \"./llama.cpp-$LLAMA_CPP_VERSION\"",
```

Comment (hiento09) [reply extensions/huggingface-extension/package.json:12]: Hi @Helloyunho, our CI test on linux has an error due to the differences in how the tar command operates between UNIX-like systems such as macOS and Linux. In macOS, tar can automatically interpret and use wildcards (like * in */convert.py) without needing additional options. However, on many versions of Linux, you need to add the --wildcards flag to the tar command to allow pattern matching.

Comment (hiento09) [reply extensions/huggingface-extension/package.json:12]: Tested on our ubuntu CI agent 
image

Comment (Helloyunho) [reply extensions/huggingface-extension/package.json:12]: Ah, another difference between bsdtar and GNU tar...

I'll use your suggestion

**Review by louis-jan — APPROVED** [review]
LGTM

**Comment by Van-QA** [comment]
linking this to https://github.com/janhq/jan/issues/1056

**Pull request merged**